mayastor/src/core/bio.rs

```diff
@@ -0,0 +1,316 @@
+            -7 => Self::Aborted,
+            -6 => Self::FirstFusedFailed,
+            -5 => Self::MisCompared,
+            -4 => Self::NoMemory,
+            -3 => Self::ScsiError,
+            -2 => Self::NvmeError,
+            -1 => Self::Failed,
+            0 => Self::Pending,
+            1 => Self::Success,
+            _ => panic!("invalid status code"),
+        }
+    }
+}
+
+impl From<IoStatus> for i32 {
+    fn from(i: IoStatus) -> Self {
+        match i {
+            IoStatus::Aborted => -7,
+            IoStatus::FirstFusedFailed => -6,
+            IoStatus::MisCompared => -5,
+            IoStatus::NoMemory => -4,
+            IoStatus::ScsiError => -3,
+            IoStatus::NvmeError => -2,
+            IoStatus::Failed => -1,
+            IoStatus::Pending => 0,
+            IoStatus::Success => 1,
+        }
+    }
+}
+
+impl From<i8> for IoStatus {
+    fn from(status: i8) -> Self {
+        (status as i32).into()
+    }
+}
+
+#[derive(Clone)]
+#[repr(transparent)]
+pub struct Bio(NonNull<spdk_bdev_io>);
+
+impl From<*mut c_void> for Bio {
+    fn from(io: *mut c_void) -> Self {
+        Bio(NonNull::new(io as *mut spdk_bdev_io).expect("null ptr"))
+    }
+}
+
+impl From<*mut spdk_bdev_io> for Bio {
+    fn from(io: *mut spdk_bdev_io) -> Self {
+        Bio(NonNull::new(io as *mut spdk_bdev_io).expect("null ptr"))
+    }
+}
+
+impl Bio {
+    /// obtain the Bdev this IO is associated with
+    pub(crate) fn bdev(&self) -> Bdev {
+        unsafe { Bdev::from(self.0.as_ref().bdev) }
+    }
+
+    #[inline]
+    pub(crate) fn ok(&self) {
+        unsafe {
+            spdk_bdev_io_complete(self.0.as_ptr(), IoStatus::Success.into())
+        }
+    }
+
+    /// mark the IO as failed
+    #[inline]
+    pub(crate) fn fail(&self) {
+        unsafe {
+            spdk_bdev_io_complete(self.0.as_ptr(), IoStatus::Failed.into())
+        }
+    }
+
+    /// mark the IO as impossible to submit due to a memory constraint
+    #[inline]
+    pub(crate) fn no_mem(&self) {
+        unsafe {
+            spdk_bdev_io_complete(self.0.as_ptr(), IoStatus::NoMemory.into())
+        }
+    }
+
+    /// assess the IO if we need to mark it failed or ok.
+    /// obtain the Nexus struct embedded within the bdev
+    pub(crate) fn nexus_as_ref(&self) -> &Nexus {
+        let b = self.bdev();
+        assert_eq!(b.product_name(), NEXUS_PRODUCT_ID);
+        unsafe { Nexus::from_raw((*b.as_ptr()).ctxt) }
+    }
+
+    /// get the context specifics of this IO
+    #[inline]
+    pub(crate) fn specific_as_mut<T>(&mut self) -> &mut T {
+        unsafe { &mut *(self.0.as_mut().driver_ctx.as_mut_ptr() as *mut T) }
+    }
+
+    /// get the IO context specific information for this IO
+    #[inline]
+    pub(crate) fn specific<T>(&self) -> &T {
+        unsafe { &*(self.0.as_ref().driver_ctx.as_ptr() as *const T) }
+    }
+
+    /// get a raw pointer to the base of the iov
+    #[inline]
+    pub(crate) fn iovs(&self) -> *mut spdk_sys::iovec {
```

Comment:
This function is probably unsafe?

Reply:
return a *mut T is not unsafe, dereferencing it is AFAIK?

Reply:
Hmm according to https://doc.rust-lang.org/book/ch19-01-unsafe-rust.html#dereferencing-a-raw-pointer you are indeed correct.